ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImpleHolder.java

```diff
@@ -0,0 +1,87 @@
+ * Copyright 2016, Red Hat Middleware LLC, and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.arjuna.ats.internal.jta.transaction.arjunacore.jca;
+
+import com.arjuna.ats.internal.jta.transaction.arjunacore.subordinate.jca.TransactionImple;
+import com.arjuna.ats.jta.logging.jtaLogger;
+
+import javax.transaction.xa.XAException;
+
+/**
+ * This TransactionImple holder is a performance optimisation to avoid having to synchronize on the imported
+ * transaction map in {@link com.arjuna.ats.internal.jta.transaction.arjunacore.jca.TransactionImporterImple}
+ */
+public class TransactionImpleHolder {
+    public final long MAX_MS_WAIT_TIME = 1000; // maximum no of ms to wait for the setter to be called
+
+    private volatile TransactionImple imported;
+
+    /**
+     * Wait for the current value of this holder to be updated.
+     *
+     * @return the non null value held in this holder object
+     * @throws XAException with the value XA_RETRY if the value is not updated within
+     * {@link com.arjuna.ats.internal.jta.transaction.arjunacore.jca.TransactionImpleHolder#MAX_MS_WAIT_TIME}
+     * milliseconds or if the wait is interrupted.
+     */
+    public TransactionImple getImported() throws XAException {
+        TransactionImple imported = this.imported;
+
+        if (imported != null) {
+            return imported;
+        }
+
+        long maxMillis = System.currentTimeMillis() + MAX_MS_WAIT_TIME;
+
+        synchronized (this) {
+            for (;;) {
+                imported = this.imported;
+
+                if (imported != null) {
+                    return imported;
+                }
+
+                if (maxMillis < System.currentTimeMillis()) {
+                    jtaLogger.i18NLogger.warn_transaction_import_timedout();
+                    // throw XA_RETRY since the wait may have timed out because the system is loaded
+                    throw new XAException(XAException.XA_RETRY);
+                }
+
+                try {
+                    wait(1000); // timed wait to protect against the setter never getting called
```

Comment:
Well that make sense. @tomjenkinson are you okay with just making sure that our code will allways invoke the setter (I ask because it means keeping the unbounded for (;;) loop)?